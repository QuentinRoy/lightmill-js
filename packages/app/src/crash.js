/* eslint-disable no-param-reassign */

import { XP_APP_BASE_CLASS_NAME } from './commons';
import crashTemplate from './views/crash.pug';
import './views/crash.scss';

/**
 * Show the crash view.
 * @param  {HTMLElement} node The node where to mount the view.
 * @param  {String} message The error message.
 * @param  {Error} [error] The error that has been raised.
 * @param  {Object} [run] The current run.
 * @returns {undefined}
 */
export default (node, message, error, run) => {
  node.innerHTML = crashTemplate({
    run,
    message,
    stack: error && error.stack
  });
  const detailsButton = node.querySelector('.lightmill-details-button');
  detailsButton.addEventListener('click', evt => {
    evt.preventDefault();
    node
      .querySelector(`.${XP_APP_BASE_CLASS_NAME}`)
      .classList.toggle('lightmill-with-details');
  });
  return {
    remove() {
      node.innerHTML = '';
    }
  };
};
