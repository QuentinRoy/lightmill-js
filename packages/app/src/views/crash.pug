.lightmill-base-app.lightmill-base-app-crash

  .lightmill-banner
    h1
      | Something went wrong

    div
      | Please call your experimenter.
    
  .lightmill-details
    h3
      | Run #{run ? run.id : 'unknown'}
    h4.lightmill-message= message
    if stack
      pre.lightmill-stack= stack
  
  a.lightmill-details-button(href='#')
    | show details
