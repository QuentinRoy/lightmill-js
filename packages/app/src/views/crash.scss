@import './common';

$color: #fff;
$main-background-color: #c90000;
$banner-background-color: rgba(0, 0, 0, .1);
$details-color: #820a0a;
$details-border-color: #5e0606;
$details-background-color: #fff;
$detail-button-color: #000;

.#{$app-class}-crash {
  background-color: $main-background-color;
  text-align: center;
  color: $color;

  .#{$class-prefix}banner {
    margin-top: 2em;
    background-color: $banner-background-color;
    padding: 1em;
    font-size: 2em;

    h1 {
      margin-top: 0;
    }
  }

  .#{$class-prefix}details {
    display: none;
    margin-top: 5em;
    margin-right: 10%;
    margin-left: 10%;
    border-width: 1px;
    border-style: solid;
    border-color: $details-border-color;
    background-color: $details-background-color;
    padding: 1em 2em;
    text-align: left;
    color: $details-color;
    overflow-x: scroll;
  }

  .#{$class-prefix}details-button {
    display: block;
    margin-top: 6em;
    color: $detail-button-color;
  }

  &.#{$class-prefix}with-details {
    .#{$class-prefix}details-button {
      display: none;
    }

    .#{$class-prefix}details {
      display: inherit;
    }
  }

}
