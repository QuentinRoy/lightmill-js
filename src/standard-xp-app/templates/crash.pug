.xp-app-base.xp-app-base-crash

  .banner
    h1
      | Something went wrong

    div
      | Please call your experimenter.
    
  .details
    h3
      | Run #{run ? run.id : 'unknown'}
    h4.message= message
    if stack
      pre.stack= stack
  
  a.details-button(href='#')
    | show details
