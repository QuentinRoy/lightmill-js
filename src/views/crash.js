import template from './crash.pug';

export default class CrashView {
  constructor(parentDiv) {
    this.parentDiv = parentDiv;
    this.template = template;
  }
  show(run, message, error) {
    const div = document.createElement('div');
    div.innerHTML = this.template({ run, message, stack: error.stack });
    const detailsButton = div.querySelector('.details-button');
    detailsButton.addEventListener('click', (evt) => {
      evt.preventDefault();
      div.querySelector('.details').style.display = 'inherit';
      detailsButton.style.display = 'none';
    });
    this.parentDiv.appendChild(div);
  }
}
