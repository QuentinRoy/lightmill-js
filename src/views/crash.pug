.full-parent(style=`color:white; background-color: #c90000; text-align: center;`)

  div(style={ 'font-size': '2em', 'padding': '1em', 'background-color': 'rgba(0,0,0,.1)', 'margin-top': '2em' })
    h1(style='margin-top: 0')
      | Something went wrong
      
    div
      | Please call your experimenter.
    
  .details(style={
    'margin-left': '10%',
    'margin-right': '10%',
    'margin-top': '5em',
    'border-color': '#5e0606',
    'border-width': '1px',
    'border-style': 'solid',
    'background-color': '#fff',
    'color': '#820a0a',
    'padding': '1em 2em',
    'text-align': 'left',
    'display': 'none',
    'overflow-x': 'scroll'
  })
    h3
      | Run #{run ? run.id : 'unknown'}
    h4.message= message
    if stack
      pre.stack= stack
  
  a.details-button(href='#', style={
      'display': 'block',
      'margin-top': '6em',
      'color': 'black'
    })
    | show details
