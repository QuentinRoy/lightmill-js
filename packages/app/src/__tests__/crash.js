/* globals document */

import simulant from 'simulant';
import crash from '../crash';

let node;
let err;
beforeEach(() => {
  node = document.createElement('div');
  node.innerHTML = '<div>Foo</div>';
  err = new Error('error message');
  err.stack = ['err stack'];
});

describe('Waiting view', () => {
  it('can be shown', () => {
    crash(node, 'message', err);
    expect(node).toMatchSnapshot();
  });
  it('can be removed', () => {
    const { remove } = crash(node, 'message', err);
    remove();
    expect(node).toMatchSnapshot();
  });
  it('can display additional details on the error', () => {
    crash(node, 'message', err);
    simulant.fire(node.querySelector('.lightmill-details-button'), 'click');
    expect(node).toMatchSnapshot();
  });
});
