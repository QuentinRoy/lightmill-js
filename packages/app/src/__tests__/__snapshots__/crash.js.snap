// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Waiting view can be removed 1`] = `<div />`;

exports[`Waiting view can be shown 1`] = `
<div>
  <div
    class="xp-app-base xp-app-base-crash"
  >
    <div
      class="banner"
    >
      <h1>
        Something went wrong
      </h1>
      <div>
        Please call your experimenter.
      </div>
    </div>
    <div
      class="details"
    >
      <h3>
        Run unknown
      </h3>
      <h4
        class="message"
      >
        message
      </h4>
      <pre
        class="stack"
      >
        err stack
      </pre>
    </div>
    <a
      class="details-button"
      href="#"
    >
      show details
    </a>
  </div>
</div>
`;

exports[`Waiting view can display additional details on the error 1`] = `
<div>
  <div
    class="xp-app-base xp-app-base-crash with-details"
  >
    <div
      class="banner"
    >
      <h1>
        Something went wrong
      </h1>
      <div>
        Please call your experimenter.
      </div>
    </div>
    <div
      class="details"
    >
      <h3>
        Run unknown
      </h3>
      <h4
        class="message"
      >
        message
      </h4>
      <pre
        class="stack"
      >
        err stack
      </pre>
    </div>
    <a
      class="details-button"
      href="#"
    >
      show details
    </a>
  </div>
</div>
`;
